stdlib/Dates/src/arithmetic.jl

```diff
@@ -95,5 +95,17 @@ end
 (-)(x::AbstractRange{T}, y::AbstractRange{T}) where {T<:TimeType} = Vector(x) - Vector(y)
 
 # Allow dates, times, and time zones to broadcast as unwrapped scalars
-Base.Broadcast.broadcastable(x::AbstractTime) = Ref(x)
-Base.Broadcast.broadcastable(x::TimeZone) = Ref(x)
+broadcastable(x::AbstractTime) = Ref(x)
+broadcastable(x::TimeZone) = Ref(x)
+
+function broadcasted(::typeof(+), r::StepRange{T}, p::Period) where {T<:TimeType}
+    StepRange{T}(r.start + p, r.step, r.stop + p)
```

Comment:
This is a little scary to me because Date ranges are strange.  For example:

```
julia> collect(Date(2020,10,31):Month(1):Date(2021))
3-element Array{Date,1}:
 2020-10-31
 2020-11-30
 2020-12-31
```

This range represents the last day of the month.  One day prior to the last day of the month seems like it'd be:

```Julia
julia> collect(Date(2020,10,31):Month(1):Date(2021)) .- Day(1)
3-element Array{Date,1}:
 2020-10-30
 2020-11-29
 2020-12-30
```

But with this definition, we'd bring that subtraction inside the step range and get the 30th day of the month:

```Julia
julia> collect(Date(2020,10,31) - Day(1):Month(1):Date(2021) - Day(1))
3-element Array{Date,1}:
 2020-10-30
 2020-11-30
 2020-12-30
```

Reply:
Thanks for bringing this up. I think I'm personally okay with having slight differences between ranges and vectors as you can always use `collect` to get the original behavior.

That said I can understand wanting the behavior to be consistent between vectors and ranges so we can always make arithmetic with `Year`s and `Month`s return vectors which are the only periods for `Date` and `DateTime` where this inconsistency should occur. (there maybe some with `Week` I'll check yet)

Reply:
One other point to bring up is that with `ZonedDateTime`s having the behavior be different between ranges and vectors is useful as you can create ranges that can't be represented as vectors:
```julia
julia> using Dates, TimeZones

julia> r = ZonedDateTime(2015,3,5,2,tz"America/New_York"):Day(1):ZonedDateTime(2015,3,10,tz"America/New_York")
ZonedDateTime(2015, 3, 5, 2, tz"America/New_York"):Day(1):ZonedDateTime(2015, 3, 9, 2, tz"America/New_York")

julia> collect(r)
ERROR: NonExistentTimeError: Local DateTime 2015-03-08T02:00:00 does not exist within America/New_York
```

Reply:
I verified that `Week` cannot have this kind of differences as a week is simply defined as 7 days (I wasn't sure if an ISO week-numbering would come into play here).